kubeflow/trainer/types/types.py

```diff
@@ -238,3 +242,46 @@ class Initializer:
 
     dataset: Optional[HuggingFaceDatasetInitializer] = None
     model: Optional[HuggingFaceModelInitializer] = None
+
+
+# local execution types
+
+@dataclass
+class LocalRuntime(Runtime):
+    create_venv: Optional[bool] = True
+    command: List[str] = field(default_factory=list)
+    python_path: Optional[str] = sys.executable
+    execution_dir: Optional[str] = None
+
+    def get_executable_command(self) -> str:
+        venv_path = Path(self.execution_dir)
+        command_str = " ".join(self.command).lstrip()
+        if self.create_venv:
+            if os.name == 'nt':
+                # Windows
+                command_exe = venv_path / "Scripts" / command_str
+            else:
+                # Unix / macOS
+                command_exe = venv_path / "bin" / command_str
+        else:
+            command_exe = command_str
+
+        # @szaher need to make sure venv is created before this check
+        # if not command_exe.exists():
+        #     raise FileNotFoundError(
+        #     f"Python executable not found in virtualenv at: {command_exe}")
+
+        return str(command_exe)
+
+
+@dataclass
+class LocalTrainJob(TrainJob):
+    job: LocalJob = None
+
+
+# Training Backends Types
+# this can be simplified if we drop python3.9 support as follows
+RuntimeList = Union[List[Runtime], List[LocalRuntime]]
+TrainingRuntime = Union[Runtime, LocalRuntime]
+TrainJobLike = Union[TrainJob, LocalTrainJob]
```

Comment:
I think, we should keep user-facing types consistent across all backends.
That will make APIs easier to use, and give users ability to quickly switch between backends without changing their code.

If we require additional parameters for other backends, we can keep them internal.

Reply:
@szaher We need to make sure we cover it in this PR, since we don't want to introduce breaking changes to the SDK. Which means public APIs should always return `Runtime  and TrainJob` object types.